XMPP HTTP upload

flowolf opened the pull request:
## Description:

Add support for sending files through Jabber and HTTP upload (XEP_0363)
can load a file from a local file or a remote URL. e.g. images. 🖼

**Related issue (if applicable):** none

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6788

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example configuration.yaml entry
notify:
  - name: NOTIFIER_NAME  # e.g. jabber
    platform: xmpp
    sender: YOUR_JID
    password: YOUR_JABBER_ACCOUNT_PASSWORD
    recipient: YOUR_RECIPIENT
    tls: true
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

flowolf committed:
notify.xmpp: first working http upload

flowolf committed:
extension guessing for upload

flowolf committed:
docstrings, flake8, pylint

flowolf committed:
hass.async_add_executor_job(...)

flowolf committed:
catch more errors, allow unverified SSL request

allow user to specify unverified SSL request to URL
cleaner code
catch more exceptions

flowolf committed:
pylint

fabaff commented:
It seems that all servers that I'm using do not support working with files/images.

Service call run without error.

```json
{
  "message": "{{ now() }}",
  "title": "The sun is {% if is_state('sun.sun', 'above_horizon') %}up{% else %}down{% endif %}!",
  "data": {
        "url": "https://github.com/home-assistant/home-assistant.io/raw/next/source/images/favicon-192x192.png"
  }
}
```

According to the code the sending of the message and the file are two tasks. In my case the text message alone works but not if there is a file attached (as in the service call above). Shouldn't at least the message get through?

fabaff commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,20 +103,204 @@ def __init__(self):
-            self.disconnect(wait=True)
+
+        # pylint: disable=no-self-use
+        def get_extension(self, content_type):
```
This looks like a task for `python-magic` or a similar module.

flowolf replied:
I'm actually trying to do the opposite of `magic`. 

A HTTP request answers with a special header `Content-Type`. I use that to 'guess' a file extension that can be used. We don't get a filename, except if specified with `path`, so the upload has to be made to a 'fictional' file name. This is `upload.EXTENSION`. We can do it without, but then a mobile client such as Conversations, does not know what to use to open it. That's why we 'guess' an appropriate extension.
This is not used when a local file is specified...

apollo13 replied:
Can't the `mimetypes` module do what you want? it has a function to guess the filetype based on a mime type…

```
In [2]: mimetypes.guess_extension('text/calendar')                                                                                                                                                    
Out[2]: '.ics'
```

flowolf replied:
YES! thank you! I'll use that.
there are some differences and unavailable filetypes, but I guess we can live with that:
```
mimetypes  vs  current_dict
None .aac
None .azw
.a .bin
None .bmp
None .bz
None .bz2
None .epub
None .ico
.jpe .jpg
.mid .midi
None .midi
.m1v .mpeg
None .mpkg
None .otf
.pot .ppt
None .rar
None .ts
None .ttf
.bat .txt
None .wav
None .weba
None .webp
None .woff
None .woff2
.xlb .xls
.rdf .xml
None .3g2
```

apollo13 replied:
@flowolf Parts of your None-matches in `mimetypes` occurs because the content-type you have in your list does not match `/etc/mime.types`. Ie `application/rar` vs `application/x-rar-compressed`. `.mid` vs `.midi` can be (in most cases) attacked by `guess_all_extensions` and choossing the longest.

fabaff commented:
Uff, after 10 min there is the error:

```bash
2018-10-15 23:26:37 INFO (MainThread) [homeassistant.components.notify.xmpp] getting file from https://github.com/home-assistant/home-assistant.io/raw/next/source/images/favicon-192x192.png
2018-10-15 23:37:32 ERROR (MainThread) [homeassistant.core] Error doing job: Task was destroyed but it is pending!
2018-10-15 23:37:32 ERROR (MainThread) [homeassistant.core] Error doing job: Future exception was never retrieved
slixmpp.exceptions.IqError: <iq xml:lang="en" to="notifications@jabber.me/home-assistant" from="dict.jabber.me" type="error" id="89220499-2768-4abd-b5dd-7f3aa7c74d49-23"><query xmlns="http://jabber.org/protocol/disco#info" /><error code="501" type="cancel"><feature-not-implemented xmlns="urn:ietf:params:xml:ns:xmpp-stanzas" /><text xmlns="urn:ietf:params:xml:ns:xmpp-stanzas">The feature requested is not implemented by the recipient or server and therefore cannot be processed.</text></error></iq>
```

flowolf commented:
I'll have a look. Maybe I've to do extra service discovery earlier to see if the server supports it. Or maybe we can set a shorter timeout. I'm on it!

thundergreen commented:
Woooow...that's awesome dude! I was praying for soooo long to get this feature.please be aware that the pic has to be wrapped into oob to make it compatible with conversations and other XMPP clients

flowolf committed:
catching XMPP exceptions, timeout for requests call

removed calls for roster and presence
added timeout for upload request call
cleared up debug, info, warning messages
cleared up requests call for secure and insecure retrieval of image
catching IqError, IqTimeout, XMPPError from slixmpp
docstring updated

flowolf commented:
@fabaff I overlooked the `slixmpp.exceptions.IqError`. I'm catching that and others now.
Also I've added a timeout to the requests call (getting the external resource), defaults to 30 seconds. 
Further the code now catches errors closer to the send call, so that the text part of a message is more likely to get trough, even if the image upload fails.
The code now catches:
- wrong credentials
- missing XEP_0363 on the server side
- non-existing URL - via timeout
- upload errors
- SSL Exceptions
- file access issues
and others.

the XMPP lib checks on its own if the server supports XEP_0363, if not, an exception is caught and an error is logged.
there is a lot that can go wrong. I hope most of it is contained.

@thundergreen I know about OOB. The docs will document that the server needs to have XEP_0363 enabled: [future-docs](https://deploy-preview-6788--home-assistant-docs.netlify.com/components/notify.xmpp#jabber-text-message). Also I'm testing with Conversations. It would be great if you could test the code as well, with different clients when the beta is out; whenever this will be merged.

thundergreen commented:
@flowolf I'll test it as soon as it is out. Not clear how to attach the picture ....with data? Afaik will the lib itself take care of oob and the rest ..I just wouldn't know how to implement this. 

I'm really excited to test. Unfortunately I only use conversations...well more like a custom fork but in the core the same with all xmop standards included ..so testing would have same base.

flowolf commented:
@thundergreen the second `data` block is optional and refers to the image or file you want to send. maybe we should rename this to `file`?
the script shall just illustrate the structure of the data that is passed to the platform. @fabaff mentioned [his JSON ](https://github.com/home-assistant/home-assistant/pull/17426#issuecomment-430021851) version above. You can pass the JSON into the services developer tools for the XMPP service as `service data`.

flowolf committed:
added timout for http upload of local files

flowolf commented:
I reproduced the problem. Currently when trying to send a file through `jabber.sk`, there are issues. It seems to be a library issue with slixmpp. I opened up [an issue on their gitlab page](https://lab.louiz.org/poezio/slixmpp/issues/3420), and am waiting for a response.

I checked with @thundergreen 's ejabberd, and it worked. My prosody server works fine as well.

The behavior with jabber.sk is that slixmpp does not handle the discovery correctly, does not see or react to the upload feature, and/or waits around for responses from servers. Not sure what goes on.

flowolf committed:
timeout, mimetypes, random filenames

guessing filetypes and mimetypes with stdlib mimetypes
setting a random filename for privacy
working around slixmpp timeout issues with asyncio.wait_for

thundergreen commented:
Is it possible to push this pr toi test ?

flowolf committed:
code cleanup

flowolf committed:
added file upload for rooms/groupchats

flowolf commented:
I cleaned up the code, and added support for sending file/images to rooms.

the code is as feature complete as I can get it for HTTP-upload. 
The issues that are present with jabber.sk, are handled with timeouts and logged as errors. Users should be able to see that their Jabber server does not work as intended.

@fabaff please have another look.

thundergreen commented:
Sounds great .thanks for that.hope it will be merged into Dev branch

flowolf commented:
there is a fix of some issues on the way in slixmpp. I'll integrate those with a new version.

flowolf committed:
version bump for slixmpp to 1.4.1

added NotConnectedError, removed double catches of IqErrors
removed asyncio import

flowolf committed:
slixmpp 1.4.1 in requirements_all

flowolf committed:
added url and path templating

fabaff committed:
Minor changes

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+        async def start(self, event):
             """Start the communication and sends the message."""
-            self.get_roster()
-            self.send_presence()
```
Was removing `get_roster()` and `send_presence()` done on purpose, it seems unrelated?

flowolf replied:
It's not needed to send a message, must have been there from sample code before. adds time and possible places for error to the request.

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+                _LOGGER.error("Error reading file %s", ex)
+            except FutTimeoutError as ex:
+                _LOGGER.error("The server did not respond in time, %s", ex)
             else:
```
This seems a bit odd, could the `else` part just be moved inside the `try` block to keep things together?

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+                _LOGGER.debug("Timeout set to %ss", timeout)
+                url = await self.upload_file(timeout=timeout)
+                if url is None:
+                    raise FileUploadError("Could not upload file")
```
Raising that in `upload_file` could simplify the error handling here.

flowolf replied:
yes, that cleans up the code, will do that

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+                _LOGGER.debug(
+                    "Got url template: %s", data.get(ATTR_URL_TEMPLATE))
+                templ = template_helper.Template(
+                    data.get(ATTR_URL_TEMPLATE), hass)
```
We can use `data[ATTR_URL_TEMPLATE]` here because we already tested it in the `if`.

(And the same applies to the branches below)

flowolf replied:
yes, will do that.

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+            _LOGGER.info("Getting file from %s", url)
+
+            def get_url(url):
+                return requests.get(
```
Since you are moving to async, it is a bit odd to introduce `requests` usage.

flowolf replied:
yes, I know. Not that familiar with aiohttp yet. requests was my goto. might patch that when I'm up to speed with it.

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+            """Upload a file from a local file path via XEP_0363."""
+            _LOGGER.info('Uploading file from path, %s ...', path)
+
+            with open(path, 'rb') as upfile:
```
Maybe we should check for whitelisted paths with `hass.config.is_allowed_path(path)`?

I am not sure when that is supposed to be used.

flowolf replied:
YES! thank you. good catch. We should enforce the whitelist here.

amelchio commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+        def get_random_filename(self, filename):
+            """Return a random filename, leaving the extension intact."""
+            if '.' in filename:
+                extension = filename.split('.')[-1]
```
This could fail for strings like `/home/user/data/../file` which is maybe not completely unlikely with templates in use(?)

flowolf replied:
yes, will fix.

amelchio reviewed (approved):
Looks good! I have added a few comments that you may want to address but nothing big.

flowolf commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -59,18 +77,26 @@ def __init__(self, sender, resource, password,
         title = kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT)
-        data = '{}: {}'.format(title, message) if title else message
+        text = '{}: {}'.format(title, message) if title else message
+        data = None or kwargs.get(ATTR_DATA)
```
yes, you are right. this could be `data = kwargs.get(ATTR_DATA)

fabaff commented on homeassistant/components/notify/xmpp.py:
```diff
@@ -93,29 +118,217 @@ def __init__(self):
+            """
+            _LOGGER.info("Getting file from %s", url)
+
+            def get_url(url):
```
Docstring is missing. Sorry, didn't see that earlier.

flowolf replied:
ACK. will fix

flowolf committed:
fixed review requests

fixed possible path issue for foo/../bar/ paths
fixed possible access for non-whitelisted files
fixed None or X
fixed try-else block, moved else block into try
fixed raising error in upload_file if url is None
fixed using data.get after it's already been checked
fixed added docstring for tiny get_url function

flowolf commented:
@amelchio @fabaff thanks for the reviews!
